baselines/tamuna/tamuna/strategy.py

```diff
@@ -0,0 +1,180 @@
+"""Tamuna and FedAvg strategies."""
+
+from typing import Callable, List
+
+import flwr.common
+import numpy as np
+import torch
+from flwr.common import (
+    FitIns,
+    NDArrays,
+    Scalar,
+    ndarrays_to_parameters,
+    parameters_to_ndarrays,
+)
+from flwr.server.strategy import Strategy
+
+from tamuna.models import Net
+
+
+def aggregate(weights: List[NDArrays], sparsity: float) -> NDArrays:
+    """Compute average of the clients' weights."""
+    averaged_weights = [
+        np.sum(layer_updates, axis=0) / sparsity for layer_updates in zip(*weights)
+    ]
+    return averaged_weights
+
+
+def create_pattern(dim: int, cohort_size: int, sparsity: int):
+    """Create compression pattern."""
+    pattern = torch.zeros(size=(dim, cohort_size))
+    if dim >= cohort_size / sparsity:
+        k = 0
+        for i in range(dim):
+            for _ in range(sparsity):
+                pattern[i, k] = 1
+                k = (k + 1) % cohort_size
+    else:
+        k = 0
+        for _ in range(sparsity):
+            for i in range(dim):
+                pattern[i, k] = 1
+                k += 1
+
+    return pattern
+
+
+def shuffle_columns(pattern: torch.Tensor):
+    """Shuffle the columns of the compression pattern."""
+    pattern = pattern[:, torch.randperm(pattern.size()[1])]
+    return pattern
+
+
+class TamunaStrategy(Strategy):
```

Comment:
Following the logic in my comment on the `readme.md`, you could also make the `Tamuna` strategy, to inherit from `FedAvg` and only override those methods that are different from plain `FedAvg`.  For instance `initialize_parameters()` is never used...

Reply:
Are you referring to `initialize_parameters()` of `TamunaStrategy`? This method is mainly used to initialize the compression pattern for clients and is called and executed once at the beginning. I think I can't reuse any of the `FedAvg` methods in `Tamuna` unfortunately. One of the reasons is that compression needs to be called after training on clients (cannot be called at evaluate). I could theoretically use evaluate for clients to do the compression, but that seems very forced and not clean.

Reply:
`configure_fit` needs to be overridden to add shuffling of the compression pattern
`aggregate_fit` different way of aggregating, needs to be overriden
`initialize_parameters` is different, 
`configure_evaluate` and `aggregate_evaluate` are not used in `Tamuna`

I see two options: extend `FedAvg` and override all of the methods or write `Tamuna` as a separate strategy (extending only base `Strategy`). I think second is better in this case.

Reply:
ok. let's leave the `Tamuna` strategy as is.